<a href="https://colab.research.google.com/github/Microblast56/Flyrank_ml_Internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Microblast56/Flyrank_ml_Internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

## 1. My lane (or freestyle) and why

**Lane: Refresh / Content Opportunity Scoring**

I want to investigate how content and search-performance signals can be used
to prioritize pages for human review. The goal is to produce a ranked list of
pages that may deserve actions such as refresh, expansion, protection,
pruning, or monitoring. I chose this lane because the starter notebooks showed
that several observable signals can be combined to rank pages, while a simple
fixed rule may not capture every useful pattern.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*


The decision is which content pages should be reviewed first by an SEO or
content editor.

The person using the output would review the highest-priority pages and decide
whether to refresh, expand, protect, prune, or monitor them.

A wrong recommendation can waste limited editorial time by directing attention
to a page that does not need intervention. Another costly error is failing to
surface an important page that should have been reviewed. The project will
therefore focus on useful prioritization rather than treating a score as a
guarantee.

Data and ML may help because page performance can depend on several observable
signals at the same time, including visibility, clicks, CTR, position,
freshness, age, engagement, and content characteristics. A fixed rule may not
capture all of these combinations.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [ ]:
import os
import pandas as pd
import subprocess
import sys

REPO_DIR = "/content/flyrank-ml-internship-starter"

# Clone the public starter repo if it is not already present
if not os.path.isdir(REPO_DIR):
    subprocess.run(
        [
            "git",
            "clone",
            "--depth",
            "1",
            "https://github.com/flyrank-bih/flyrank-ml-internship-starter",
            REPO_DIR,
        ],
        check=True,
    )

# Load the starter dataset
csv_path = os.path.join(
    REPO_DIR,
    "data",
    "raw",
    "content_refresh_anonymized.csv"
)

df = pd.read_csv(csv_path)

# Starter proxy label
declining = df["trend_direction"].str.lower().eq("down")

# Supporting numbers for the research question
total_pages = len(df)
declining_pages = declining.sum()
declining_rate = declining.mean()
visible_declining_pages = (
    declining & (df["impressions_90d"] >= 500)
).sum()

print(f"Total pages: {total_pages:,}")
print(f"Declining pages: {declining_pages:,}")
print(f"Declining rate: {declining_rate:.1%}")
print(f"Declining pages with >=500 impressions: {visible_declining_pages:,}")

Total pages: 30,000
Declining pages: 16,262
Declining rate: 54.2%
Declining pages with >=500 impressions: 9,961


## 3. Quick look at the data (2-3 real numbers)

The starter dataset contains **30,000 content pages**. Using the starter
proxy label, **16,262 pages (54.2%)** are classified as declining.

Among those declining pages, **9,961 have at least 500 impressions** in the
90-day window. This suggests that there is a substantial set of visible pages
where prioritizing human review could be useful.

These numbers make the refresh and content-opportunity problem worth
investigating further, while the next stages will test whether better ranking
methods can improve prioritization.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

This work can identify observed relationships between content and search
signals and help create a directional ranking of pages for human review. It
can provide decision-support by showing which pages appear more worthy of
attention based on the signals and validation methods used.

I cannot claim that the analysis proves why Google ranks a page in a certain
way, that a recommended refresh will definitely improve performance, or that
the model predicts Google's algorithm.

I also cannot treat the starter proxy label as proof of a future decline. Any
future model should use clearly defined future outcomes and leakage-safe
validation before stronger predictive claims are made.

## Self-check

Before you submit, confirm each line honestly:

- [✔] Every section above is filled — markdown thinking AND the code that backs it
- [✔] The notebook runs top to bottom with no errors (Runtime → Run all)
- [✔] No client names, URLs, or private queries anywhere
- [✔] My claims use careful words: observed, measured, directional, decision-support
- [✔] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.